flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/maintenance/api/TestExpireSnapshots.java

```diff
@@ -0,0 +1,235 @@
+  }
+
+  @Test
+  void testExpireSnapshots() throws Exception {
+    insert(table, 1, "a");
+    insert(table, 2, "b");
+    insert(table, 3, "c");
+    insert(table, 4, "d");
+
+    Set<Snapshot> snapshots = Sets.newHashSet(table.snapshots());
+    assertThat(snapshots).hasSize(4);
+
+    ExpireSnapshots.builder()
+        .parallelism(1)
+        .planningWorkerPoolSize(2)
+        .deleteBatchSize(3)
+        .deleteParallelism(1)
+        .maxSnapshotAge(Duration.ZERO)
+        .retainLast(1)
+        .uidSuffix(UID_SUFFIX)
+        .append(
+            infra.triggerStream(),
+            0,
+            DUMMY_NAME,
+            tableLoader(),
+            "OTHER",
+            StreamGraphGenerator.DEFAULT_SLOT_SHARING_GROUP,
+            1)
+        .sinkTo(infra.sink());
+
+    runAndWaitForSuccess(
+        infra.env(), infra.source(), infra.sink(), () -> checkDeleteFinished(3L), table);
+
+    table.refresh();
+    assertThat(Sets.newHashSet(table.snapshots())).hasSize(1);
+    // Check that the table data not changed
+    SimpleDataUtil.assertTableRecords(
+        table,
+        ImmutableList.of(
+            createRecord(1, "a"),
+            createRecord(2, "b"),
+            createRecord(3, "c"),
+            createRecord(4, "d")));
+  }
+
+  @Test
+  void testFailure() throws Exception {
+    insert(table, 1, "a");
+    insert(table, 2, "b");
+
+    SerializableTable serializableTable = (SerializableTable) SerializableTable.copyOf(table);
+
+    ExpireSnapshots.builder()
+        .append(
+            infra.triggerStream(),
+            0,
+            DUMMY_NAME,
+            tableLoader(),
+            UID_SUFFIX,
+            StreamGraphGenerator.DEFAULT_SLOT_SHARING_GROUP,
+            1)
+        .sinkTo(infra.sink());
+
+    JobClient jobClient = null;
+    try {
+      jobClient = infra.env().executeAsync();
+
+      // Do a single task run
+      long time = System.currentTimeMillis();
+      infra
+          .source()
+          .sendRecord(Trigger.create(time, serializableTable, 1), System.currentTimeMillis());
+
+      // First successful run (ensure that the operators are loaded/opened etc.)
+      assertThat(infra.sink().poll(Duration.ofSeconds(5)).success()).isTrue();
+
+      // Drop the table, so it will cause an exception
+      dropTable();
+
+      // Failed run
+      infra.source().sendRecord(Trigger.create(time + 1, serializableTable, 1), time + 1);
+
+      assertThat(infra.sink().poll(Duration.ofSeconds(5)).success()).isFalse();
+    } finally {
+      closeJobClient(jobClient);
+    }
+
+    // Check the metrics
+    MetricsReporterFactoryForTests.assertCounters(
+        new ImmutableMap.Builder<String, Long>()
+            .put(
+                DELETE_FILES_OPERATOR_NAME + "." + DUMMY_NAME + "." + DELETE_FILE_FAILED_COUNTER,
+                0L)
+            .put(
+                DELETE_FILES_OPERATOR_NAME + "." + DUMMY_NAME + "." + DELETE_FILE_SUCCEEDED_COUNTER,
+                0L)
```

Comment:
there was a successful task run in line 117-124. why is the success counter still zero?

Reply:
There was no actual deletes at that time. No files to remove, so the counters are not incremented

Reply:
oh. that is because we didn't set the maxAge or maxSnapshots in the builder. maybe add a comment in line 133 or line 119. maybe sth like `there are no expired snapshots or data files because ExpireSnapshots has no max age of number of snapshots set`.

Reply:
Done